frontend/plugincompatibility.lua

```diff
@@ -0,0 +1,423 @@
+    end
+
+    return override.action
+end
+
+--- Set the load override for a plugin.
+-- @tparam table G_reader_settings
+-- @tparam string plugin_name
+-- @tparam string plugin_version
+-- @tparam string action "always", "never", or "load-once"
+function PluginCompatibility.setLoadOverride(G_reader_settings, plugin_name, plugin_version, action)
+    local overrides = G_reader_settings:readSetting("plugin_load_overrides") or {}
+    local koreader_version = Version:getShortVersion()
+
+    if action == nil or action == "ask" then
+        -- Remove override
+        overrides[plugin_name] = nil
+    else
+        overrides[plugin_name] = {
+            action = action,
+            version = plugin_version,
+            koreader_version = koreader_version,
+        }
+    end
+
+    G_reader_settings:saveSetting("plugin_load_overrides", overrides)
+end
+
+--- Clear the load-once override after it has been used.
+-- @tparam table G_reader_settings
+-- @tparam string plugin_name
+function PluginCompatibility.clearLoadOnceOverride(G_reader_settings, plugin_name)
+    local overrides = G_reader_settings:readSetting("plugin_load_overrides") or {}
+    local override = overrides[plugin_name]
+
+    if override and override.action == "load-once" then
+        overrides[plugin_name] = nil
+        G_reader_settings:saveSetting("plugin_load_overrides", overrides)
+    end
+end
+
+--- Determine if a plugin should be loaded based on compatibility and overrides.
+-- @tparam table G_reader_settings
+-- @tparam table plugin_meta The plugin's metadata
+-- @tparam string plugin_name
+-- @treturn boolean true if should load, false otherwise
+-- @treturn string|nil reason for not loading or nil
+-- @treturn string|nil incompatibility message or nil
+-- @treturn boolean true if user should be prompted
+function PluginCompatibility.shouldLoadPlugin(G_reader_settings, plugin_meta, plugin_name)
+    local is_compatible, reason, message = PluginCompatibility.checkCompatibility(plugin_meta)
+
+    if is_compatible then
+        -- Plugin is compatible, load it
+        return true, nil, nil, false
+    end
+
+    -- Plugin is incompatible, check for overrides
+    local plugin_version = plugin_meta and plugin_meta.version or "unknown"
+    local override = PluginCompatibility.getLoadOverride(G_reader_settings, plugin_name, plugin_version)
+
+    if override == "always" then
+        -- User wants to always load this plugin despite incompatibility
+        return true, nil, nil, false
+    elseif override == "never" then
+        -- User explicitly doesn't want this plugin loaded
+        return false, reason, message, false
+    elseif override == "load-once" then
+        -- User wants to load it once for testing
+        -- Clear the override so next time it won't auto-load
+        PluginCompatibility.clearLoadOnceOverride(G_reader_settings, plugin_name)
+        PluginCompatibility.removePromptedMark(G_reader_settings, plugin_name, plugin_version)
```

Comment:
I do agree with @poire-z vis-à-vis <https://github.com/koreader/koreader/pull/14655#issuecomment-3572243283> that extracting a handful of meager lines like this into their entirely separate functions doesn't aid clarity. Presumably it's primarily done for the sake of being able to test it in isolation, but as a consequence instead of having ~10 clear lines, it's ~30 unclear "overengineered" lines that require hopping back and forth. It's a trade-off of course, and one that bothers me less than it does @poire-z, but I am in the same camp nonetheless.

Besides reduced maintainability aspect, it's also a mildly suspicious pattern in the sense that it would be worse in performance-sensitive code. As such doing so unnecessarily might be a bad habit even where it makes no difference.

Reply:
Testing is one key contributor for splitting the logic into smaller units. TDD (test driven development) is one way to get things done. This allows for one to no longer care about the internals of a function that is under test once it passes them.

Nevertheless, I am fine with merging things in bigger functions if that is the preference. I would need some pointers on what you would like to see merged. Since we see things differently. 
I am also happy to just deliver a working version and you folks can take it over from there and refactor it as needed.